pkg/localstore/migration_postage_index.go

```diff
@@ -0,0 +1,135 @@
+	start := time.Now()
+
+	retrievalDataIndex, err := db.shed.NewIndex("Address->StoreTimestamp|BinID|BatchID|BatchIndex|Sig|Data", shed.IndexFuncs{
+		EncodeKey: func(fields shed.Item) (key []byte, err error) {
+			return fields.Address, nil
+		},
+		DecodeKey: func(key []byte) (e shed.Item, err error) {
+			e.Address = key
+			return e, nil
+		},
+		EncodeValue: func(fields shed.Item) (value []byte, err error) {
+			b := make([]byte, headerSize)
+			binary.BigEndian.PutUint64(b[:8], fields.BinID)
+			binary.BigEndian.PutUint64(b[8:16], uint64(fields.StoreTimestamp))
+			stamp, err := postage.NewStamp(fields.BatchID, fields.Index, fields.Timestamp, fields.Sig).MarshalBinary()
+			if err != nil {
+				return nil, err
+			}
+			copy(b[16:], stamp)
+			value = append(b, fields.Data...)
+			return value, nil
+		},
+		DecodeValue: func(keyItem shed.Item, value []byte) (e shed.Item, err error) {
+			e.StoreTimestamp = int64(binary.BigEndian.Uint64(value[8:16]))
+			e.BinID = binary.BigEndian.Uint64(value[:8])
+			stamp := new(postage.Stamp)
+			if err = stamp.UnmarshalBinary(value[16:headerSize]); err != nil {
+				return e, err
+			}
+			e.BatchID = stamp.BatchID()
+			e.Index = stamp.Index()
+			e.Timestamp = stamp.Timestamp()
+			e.Sig = stamp.Sig()
+			e.Data = value[headerSize:]
+			return e, nil
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed creating retrievalDataIndex: %w", err)
+	}
+
+	postageIndexIndex, err := db.shed.NewIndex("BatchID|BatchIndex->Hash|Timestamp", shed.IndexFuncs{
+		EncodeKey: func(fields shed.Item) (key []byte, err error) {
+			key = make([]byte, 40)
+			copy(key[:32], fields.BatchID)
+			copy(key[32:40], fields.Index)
+			return key, nil
+		},
+		DecodeKey: func(key []byte) (e shed.Item, err error) {
+			e.BatchID = key[:32]
+			e.Index = key[32:40]
+			return e, nil
+		},
+		EncodeValue: func(fields shed.Item) (value []byte, err error) {
+			value = make([]byte, 40)
+			copy(value, fields.Address)
+			copy(value[32:], fields.Timestamp)
+			return value, nil
+		},
+		DecodeValue: func(keyItem shed.Item, value []byte) (e shed.Item, err error) {
+			e.Address = value[:32]
+			e.Timestamp = value[32:]
+			return e, nil
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed creating postageIndexIndex: %w", err)
+	}
+
+	const batchSize = 1_000_000
+	var startItem *shed.Item
+	totalCleanedUpEntries := 0
+	for {
+		batch := new(leveldb.Batch)
+		currentBatchSize := 0
+		more := false
+		err := postageIndexIndex.Iterate(func(item shed.Item) (bool, error) {
+			if currentBatchSize == batchSize {
+				startItem = &item
+				more = true
+				return true, nil
+			}
+			has, err := retrievalDataIndex.Has(item)
+			if err != nil {
+				return true, err
+			}
+			if !has {
+				err := postageIndexIndex.DeleteInBatch(batch, item)
+				if err != nil {
+					return true, err
+				}
+				currentBatchSize++
+			}
+			return false, nil
+		}, &shed.IterateOptions{
+			StartFrom: startItem,
```

Comment:
starting on an empty item should be ok right?

Reply:
Yes... This same code is used in the batching logic which is tested.